pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,768 @@
+		if err != nil {
+			if errors.IsNotFound(err) {
+				reqLogger.Info("subscription :%s in namespace %s not exists, do not need update")
+				continue
+			} else {
+				return false, err
+			}
+		}
+		sub.Spec.Channel = item.Channel
+		err = c.Update(context.TODO(), sub)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// PostUpgradeVerification verify all replicasets are at expected counts and all daemonsets are at expected counts
+func PostUpgradeVerification(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+	replicaSetList := &appsv1.ReplicaSetList{}
+	err := c.List(context.TODO(), replicaSetList)
+	if err != nil {
+		return false, err
+	}
+	readyRs := 0
+	totalRs := 0
+	for _, replica := range replicaSetList.Items {
+		if strings.HasPrefix(replica.Namespace, "default") ||
+			strings.HasPrefix(replica.Namespace, "kube") ||
+			strings.HasPrefix(replica.Namespace, "openshift") {
+			totalRs = totalRs + 1
+			if replica.Status.ReadyReplicas == replica.Status.Replicas {
+				readyRs = readyRs + 1
+			}
+
+		}
+	}
+
+	if totalRs != readyRs {
+		reqLogger.Info(fmt.Sprintf("not all replicaset are ready:expected number :%v , ready number %v", len(replicaSetList.Items), readyRs))
+		return false, nil
+	}
+
+	dsList := &appsv1.DaemonSetList{}
+	err = c.List(context.TODO(), dsList)
+	if err != nil {
+		return false, err
+	}
+	readyDS := 0
+	totalDS := 0
+	for _, ds := range dsList.Items {
+		if strings.HasPrefix(ds.Namespace, "default") ||
+			strings.HasPrefix(ds.Namespace, "kube") ||
+			strings.HasPrefix(ds.Namespace, "openshift") {
+			totalDS = totalDS + 1
+			if ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
+				readyDS = readyDS + 1
+			}
+		}
+	}
+	if len(dsList.Items) != readyDS {
+		reqLogger.Info(fmt.Sprintf("not all daemonset are ready:expected number :%v , ready number %v", len(dsList.Items), readyDS))
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// TODO
+func RemoveMaintWindow(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return true, nil
+}
+
+// This perform cluster health check after upgrade
+func PostClusterHealthCheck(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return performClusterHealthCheck(c)
+}
+
+// Check whether nodes are upgraded or not
+func NodesUpgraded(c client.Client, nodeType string, reqLogger logr.Logger) (bool, error) {
+	configPool := &machineconfigapi.MachineConfigPool{}
+	err := c.Get(context.TODO(), types.NamespacedName{Name: nodeType}, configPool)
+	if err != nil {
+		return false, nil
+	}
+	//TODO send timeout alert if wait timeout
+	if configPool.Status.ReadyMachineCount != configPool.Status.UpdatedMachineCount {
+		errMsg := fmt.Sprintf("not all %s are upgraded, upgraded: %v, ready: %v", nodeType, configPool.Status.UpdatedMachineCount, configPool.Status.ReadyMachineCount)
```

Comment:
In the existing upgrade script, it compares the total number of machines in the pool (regardless of their "ready" state) vs the number updated, in order to determine if all workers are upgraded.
I mention it because I'm seeing some odd behaviour when running an upgrade with the operator in its current state. It seems to not wait at this step at all (eg I never see the "not all X are upgraded" anywhere in the logs). Not entirely sure that this check is the problem, but something doesn't look right.